File: gym/blade/Relationships.py
from typing import Dict, List, Optional


class Relationships:
    def __init__(
        self,
        hostiles: Optional[Dict[str, List[str]]] = None,
        allies: Optional[Dict[str, List[str]]] = None,
    ):
        self.hostiles: Dict[str, List[str]] = hostiles if hostiles is not None else {}
        self.allies: Dict[str, List[str]] = allies if allies is not None else {}

    def _add_hostile_one_way(self, side_id: str, hostile_id: str):
        if side_id not in self.hostiles:
            self.hostiles[side_id] = []
        if hostile_id not in self.hostiles[side_id]:
            self.hostiles[side_id].append(hostile_id)
        if side_id in self.allies:
            self.allies[side_id] = [
                id for id in self.allies[side_id] if id != hostile_id
            ]

    def _remove_hostile_one_way(self, side_id: str, hostile_id: str):
        if side_id in self.hostiles:
            self.hostiles[side_id] = [
                id for id in self.hostiles[side_id] if id != hostile_id
            ]

    def _add_ally_one_way(self, side_id: str, ally_id: str):
        if side_id not in self.allies:
            self.allies[side_id] = []
        if ally_id not in self.allies[side_id]:
            self.allies[side_id].append(ally_id)
        if side_id in self.hostiles:
            self.hostiles[side_id] = [
                id for id in self.hostiles[side_id] if id != ally_id
            ]

    def _remove_ally_one_way(self, side_id: str, ally_id: str):
        if side_id in self.allies:
            self.allies[side_id] = [
                id for id in self.allies[side_id] if id != ally_id
            ]

    def add_hostile(self, side_id: str, hostile_id: str):
        if side_id == hostile_id:
            return
        self._add_hostile_one_way(side_id, hostile_id)
        self._add_hostile_one_way(hostile_id, side_id)

    def remove_hostile(self, side_id: str, hostile_id: str):
        self._remove_hostile_one_way(side_id, hostile_id)
        self._remove_hostile_one_way(hostile_id, side_id)

    def add_ally(self, side_id: str, ally_id: str):
        if side_id == ally_id:
            return
        self._add_ally_one_way(side_id, ally_id)
        self._add_ally_one_way(ally_id, side_id)

    def remove_ally(self, side_id: str, ally_id: str):
        self._remove_ally_one_way(side_id, ally_id)
        self._remove_ally_one_way(ally_id, side_id)

    def is_ally(self, side_id: str, ally_id: str) -> bool:
        return ally_id in self.allies.get(side_id, [])

    def is_hostile(self, side_id: str, hostile_id: str) -> bool:
        return hostile_id in self.hostiles.get(side_id, [])

    def get_allies(self, side_id: str) -> List[str]:
        return self.allies.get(side_id, [])

    def get_hostiles(self, side_id: str) -> List[str]:
        return self.hostiles.get(side_id, [])

    def update_relationship(self, side_id: str, hostiles: List[str], allies: List[str]):
        next_hostiles: List[str] = []
        for hostile_id in hostiles:
            if (
                hostile_id
                and hostile_id != side_id
                and hostile_id not in next_hostiles
            ):
                next_hostiles.append(hostile_id)
        next_allies: List[str] = []
        for ally_id in allies:
            if (
                ally_id
                and ally_id != side_id
                and ally_id not in next_hostiles
                and ally_id not in next_allies
            ):
                next_allies.append(ally_id)

        prev_hostiles = list(self.hostiles.get(side_id, []))
        prev_allies = list(self.allies.get(side_id, []))

        self.hostiles[side_id] = next_hostiles
        self.allies[side_id] = next_allies

        for other_id in next_hostiles:
            if other_id not in prev_hostiles:
                self._add_hostile_one_way(other_id, side_id)
        for other_id in prev_hostiles:
            if other_id not in next_hostiles:
                self._remove_hostile_one_way(other_id, side_id)
        for other_id in next_allies:
            if other_id not in prev_allies:
                self._add_ally_one_way(other_id, side_id)
        for other_id in prev_allies:
            if other_id not in next_allies:
                self._remove_ally_one_way(other_id, side_id)

    def delete_side(self, side_id: str):
        for key in self.hostiles:
            self.hostiles[key] = [id for id in self.hostiles[key] if id != side_id]
        for key in self.allies:
            self.allies[key] = [id for id in self.allies[key] if id != side_id]
        self.hostiles.pop(side_id, None)
        self.allies.pop(side_id, None)

    def to_dict(self):
        return {
            "hostiles": self.hostiles,
            "allies": self.allies,
        }

File: client/src/game/Relationships.ts
interface IRelationships {
  hostiles?: { [sideId: string]: string[] };
  allies?: { [sideId: string]: string[] };
}

export default class Relationships {
  hostiles: { [sideId: string]: string[] };
  allies: { [sideId: string]: string[] };

  constructor(parameters: IRelationships) {
    this.hostiles = parameters.hostiles ?? {};
    this.allies = parameters.allies ?? {};
  }

  private _addHostileOneWay(sideId: string, hostileId: string) {
    if (!this.hostiles[sideId]) {
      this.hostiles[sideId] = [];
    }
    if (!this.hostiles[sideId].includes(hostileId)) {
      this.hostiles[sideId].push(hostileId);
    }
    if (this.allies[sideId]) {
      this.allies[sideId] = this.allies[sideId].filter(
        (id) => id !== hostileId
      );
    }
  }

  private _removeHostileOneWay(sideId: string, hostileId: string) {
    if (this.hostiles[sideId]) {
      this.hostiles[sideId] = this.hostiles[sideId].filter(
        (id) => id !== hostileId
      );
    }
  }

  private _addAllyOneWay(sideId: string, allyId: string) {
    if (!this.allies[sideId]) {
      this.allies[sideId] = [];
    }
    if (!this.allies[sideId].includes(allyId)) {
      this.allies[sideId].push(allyId);
    }
    if (this.hostiles[sideId]) {
      this.hostiles[sideId] = this.hostiles[sideId].filter(
        (id) => id !== allyId
      );
    }
  }

  private _removeAllyOneWay(sideId: string, allyId: string) {
    if (this.allies[sideId]) {
      this.allies[sideId] = this.allies[sideId].filter((id) => id !== allyId);
    }
  }

  addHostile(sideId: string, hostileId: string) {
    if (sideId === hostileId) return;
    this._addHostileOneWay(sideId, hostileId);
    this._addHostileOneWay(hostileId, sideId);
  }

  removeHostile(sideId: string, hostileId: string) {
    this._removeHostileOneWay(sideId, hostileId);
    this._removeHostileOneWay(hostileId, sideId);
  }

  addAlly(sideId: string, allyId: string) {
    if (sideId === allyId) return;
    this._addAllyOneWay(sideId, allyId);
    this._addAllyOneWay(allyId, sideId);
  }

  removeAlly(sideId: string, allyId: string) {
    this._removeAllyOneWay(sideId, allyId);
    this._removeAllyOneWay(allyId, sideId);
  }

  isAlly(sideId: string, allyId: string): boolean {
    return this.allies[sideId]?.includes(allyId) ?? false;
  }

  isHostile(sideId: string, hostileId: string): boolean {
    return this.hostiles[sideId]?.includes(hostileId) ?? false;
  }

  getAllies(sideId: string): string[] {
    return this.allies[sideId] ?? [];
  }

  getHostiles(sideId: string): string[] {
    return this.hostiles[sideId] ?? [];
  }

  updateRelationship(sideId: string, hostiles: string[], allies: string[]) {
    const nextHostiles = Array.from(
      new Set(hostiles.filter((id) => id && id !== sideId))
    );
    const nextAllies = Array.from(
      new Set(
        allies.filter(
          (id) => id && id !== sideId && !nextHostiles.includes(id)
        )
      )
    );
    const prevHostiles = this.hostiles[sideId] ?? [];
    const prevAllies = this.allies[sideId] ?? [];

    this.hostiles[sideId] = nextHostiles;
    this.allies[sideId] = nextAllies;

    const addedHostiles = nextHostiles.filter(
      (id) => !prevHostiles.includes(id)
    );
    const removedHostiles = prevHostiles.filter(
      (id) => !nextHostiles.includes(id)
    );
    const addedAllies = nextAllies.filter((id) => !prevAllies.includes(id));
    const removedAllies = prevAllies.filter((id) => !nextAllies.includes(id));

    addedHostiles.forEach((otherId) =>
      this._addHostileOneWay(otherId, sideId)
    );
    removedHostiles.forEach((otherId) =>
      this._removeHostileOneWay(otherId, sideId)
    );
    addedAllies.forEach((otherId) => this._addAllyOneWay(otherId, sideId));
    removedAllies.forEach((otherId) =>
      this._removeAllyOneWay(otherId, sideId)
    );
  }

  deleteSide(sideId: string) {
    Object.keys(this.hostiles).forEach((key) => {
      this.hostiles[key] = this.hostiles[key].filter((id) => id !== sideId);
    });
    Object.keys(this.allies).forEach((key) => {
      this.allies[key] = this.allies[key].filter((id) => id !== sideId);
    });
    delete this.hostiles[sideId];
    delete this.allies[sideId];
  }
}
